source/logrepl/cdc.go

```diff
@@ -0,0 +1,195 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logrepl
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/conduitio/conduit-connector-postgres/source/logrepl/internal"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/jackc/pglogrepl"
+	"github.com/jackc/pgx/v4"
+)
+
+// Config holds configuration values for our V1 Source. It is parsed from the
+// map[string]string passed to the Connector at Configure time.
+type Config struct {
+	Position        sdk.Position
+	SlotName        string
+	PublicationName string
+	TableName       string
+	KeyColumnName   string
+	Columns         []string
+}
+
+// CDCIterator listens for events from the WAL and pushes them into its buffer.
+// It iterates through that Buffer so that we have a controlled way to get 1
+// record from our CDC buffer without having to expose a loop to the main Read.
+type CDCIterator struct {
+	config   Config
+	messages chan sdk.Record
+
+	sub *internal.Subscription
+}
+
+// NewCDCIterator takes a config and returns up a new CDCIterator or returns an
+// error.
+func NewCDCIterator(ctx context.Context, conn *pgx.Conn, config Config) (*CDCIterator, error) {
+	i := &CDCIterator{
+		config:   config,
+		messages: make(chan sdk.Record),
+	}
+
+	err := i.attachSubscription(ctx, conn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup subscription: %w", err)
+	}
+
+	go i.listen(ctx)
+
+	return i, nil
+}
+
+// listen is meant to be used in a goroutine. It starts the subscription
+// passed to it and handles the subscription flush
+func (i *CDCIterator) listen(ctx context.Context) {
+	sdk.Logger(ctx).Info().
+		Str("slot", i.config.SlotName).
+		Str("publication", i.config.PublicationName).
+		Msg("starting logical replication")
+
+	err := i.sub.Start(ctx)
+	if err != nil {
+		// log it to be safe we don't miss the error, but use warn level
+		// because the error will most probably be still propagated to Conduit
+		// and might be recovered from
+		sdk.Logger(ctx).Warn().Err(err).Msg("subscription returned an error")
+	}
+}
+
+// Next returns the next record in the buffer. This is a blocking operation
+// so it should only be called if we've checked that HasNext is true or else
+// it will block until a record is inserted into the queue.
```

Comment:
Should be updated to remove references to `HasNext` since we've removed that from the interface and only have `Next` as a blocking operation now.